### PR TITLE
Specify deps with poetry / fix code rot

### PR DESCRIPTION
**Pull request by engelanna**
Using this lib in virus related research. It's very nice and minimalist, thank you for your work! :star2: :bowing_woman: 

The PR:
* explicitly pins the dependency versions, making the lib production-ready (`poetry` :heart:)
* resolves code rot that has appeared since 2018
* paves the way for publishing to PyPI soonish if the original author would be interested

Peek 2022-05-17 04-11

@clintval I'll most likely be adding simple annotation capabilities to the plot ranges, as well as adding a `.bed > pandas > genome-browser` flow. I don't need my own fork, could collab on the original directly. Your comments on any of the following are welcome:
:one: whether I should base my progress off of your [cv_update_api](https://github.com/clintval/genome-browser/tree/cv_update_api) branch instead
:two: why the right limit of the x axis seems clipped when I run the example:
Screenshot from 2022-05-17 02-51-45

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[tool.poetry]
+name = "genome-browser"
+version = "0.1.1"
+description = "A lightweight Python 3 graphing API for genomic intervals and genomic features."
+authors = ["Clint Valentine <valentine.clint@gmail.com>"]
+license = "MIT"
+
+[tool.poetry.dependencies]
+python = ">=3.7,<3.11"
+matplotlib = "^3.5.2"
```
Comment (engelanna) [inline pyproject.toml:10]: 168710268-7be6e8f8-6dca-4031-9038-ae6fe666da85

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+more-itertools = "^8.13.0"
```
Comment (engelanna) [inline pyproject.toml:11]: Screenshot from 2022-05-16 22-50-34

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+scipy = "1.7.3"
```
Comment (engelanna) [inline pyproject.toml]: Screenshot from 2022-05-16 22-52-19

scipy 1.7.3 is the latest that doesn't retire Python 3.7.

File: genome_browser/__init__.py [hunk 4]
```diff
@@ -15,7 +15,7 @@
 ASPECT = 2.8
 HSPACE = 0.05
 PADDING = 0.3
-RESOLUTION = 1e5
```
Comment (engelanna) [inline genome_browser/__init__.py:18]: Screenshot from 2022-05-17 03-45-00

File: genome_browser/__init__.py [hunk 5]
```diff
@@ -167,7 +167,7 @@ def draw(self):
                 if self.annotation is not None:
                     ax.annotate(xy=(1, 0),
                                 xycoords='axes fraction',
-                                s=self.annotation,
```
Comment (engelanna) [inline genome_browser/__init__.py:170]: Screenshot from 2022-05-17 00-59-26

File: README.md [hunk 6]
```diff
@@ -1,37 +1,39 @@
...
+## Genome Diagram
 
-As lightweight Python 3 graphing API for genomic intervals and genomic features.
+A lightweight Python 3 graphing API for genomic intervals and genomic features.
 
 ```python
->>> import numpy as np
->>> import genome_browser as gb
->>> n = 105  # Length of randoim genomic interval.
+from matplotlib import pyplot as plt
```
Comment (engelanna) [inline README.md:6]: Makes the example demonstrably work right away (1/2)

File: README.md [hunk 7]
```diff
@@ -1,37 +1,39 @@
...
+                       np.random.randint(0, 20),
+                       np.random.choice(['+', '-', 'none']),
+                       np.random.choice(['#E74C3C', '#3498DB', '0.2'])])
+g.add_track(track)
+
+# Annotate the figure with interval specific metadata. Will always appear in lower-right
+g.annotation = '{}:{:,}-{:,}'.format('chr3', 20000, 812383)
+
+fig, axes = g.draw()
+plt.show()
```
Comment (engelanna) [inline README.md:36]: Makes the example demonstrably work right away (2/2)

File: README.md [hunk 8]
```diff
@@ -1,37 +1,39 @@
...
+track = gb.Feature('Random Intervals', height_ratio=0.4)
+for _ in range(9):
+    # Feature must follow iterable as (position, width, strand, color)
+    track.add_feature([np.random.randint(0, n -15),
+                       np.random.randint(0, 20),
+                       np.random.choice(['+', '-', 'none']),
+                       np.random.choice(['#E74C3C', '#3498DB', '0.2'])])
+g.add_track(track)
+
+# Annotate the figure with interval specific metadata. Will always appear in lower-right
```
Comment (engelanna) [inline README.md:32]: It's lower-right, at least for me :woman_shrugging:

File: README.md [hunk 9]
```diff
@@ -1,37 +1,39 @@
-Genome Diagram
----
+## Genome Diagram
 
-As lightweight Python 3 graphing API for genomic intervals and genomic features.
```
Comment (engelanna) [inline README.md:4]: A typo, methinks.

File: README.md [hunk 10]
```diff
@@ -1,37 +1,39 @@
...
+A lightweight Python 3 graphing API for genomic intervals and genomic features.
 
 ```python
->>> import numpy as np
```
Comment (engelanna) [inline README.md:7]: All uncommented changes in the file are `>>>` removals.
IMHO the `>>>` get in the way under most circumstances.

File: README.md [hunk 11]
```diff
@@ -1,37 +1,39 @@
...
->>> import genome_browser as gb
->>> n = 105  # Length of randoim genomic interval.
```
Comment (engelanna) [inline README.md:9]: typo

File: README.md [hunk 12]
```diff
@@ -28,10 +29,11 @@ As lightweight Python 3 graphing API for genomic intervals and genomic features.
 >>>                        np.random.choice(['#E74C3C', '#3498DB', '0.2'])])
 >>> g.add_track(track)
 
-# Annotate the figure with interval specific metadata. Will always appear in lower-left
+# Annotate the figure with interval specific metadata. Will always appear in lower-right
 >>> g.annotation = '{}:{:,}-{:,}'.format('chr3', 20000, 812383)
 
 >>> fig, axes = g.draw()
+>>> plt.show()
```
Comment (engelanna) [inline README.md:37]: It's nice to see the example right away.

File: README.md [hunk 13]
```diff
@@ -1,12 +1,14 @@
...
----
+## Genome Diagram
 
-As lightweight Python 3 graphing API for genomic intervals and genomic features.
+A lightweight Python 3 graphing API for genomic intervals and genomic features.
 
 ```python
+>>> from matplotlib import pyplot as plt
 >>> import numpy as np
+
```
Comment (engelanna) [inline README.md:8]: External imports :point_up: separated from internal ones :point_down:

File: pyproject.toml [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+version = "0.1.1"
+description = "A lightweight Python 3 graphing API for genomic intervals and genomic features."
+authors = ["Clint Valentine <valentine.clint@gmail.com>"]
+license = "MIT"
+
+[tool.poetry.dependencies]
+python = ">=3.7,<3.11"
+matplotlib = "^3.5.2"
+more-itertools = "^8.13.0"
+scipy = "^1.7.3"
```
Comment (engelanna) [inline pyproject.toml:12]: `No module named...` + this is the latest `scipy` that doesn't retire the trusty Python 3.7.

**Comment by clintval** [comment]
Wow it is so cool you're using this prototypical (and very old) project of mine!

Ping me please when you're ready for a review!

**Comment by clintval** [comment]
Also, I had a bunch of refactor in the branch that you found that may be a good idea to pursue, but it's been neglected. What do you think of the API there? (I imagine it's very stale also)

**Comment by clintval** [comment]
If you want to go from BED, consider pulling this library, it's what I would do: https://github.com/fulcrumgenomics/pybedlite

**Comment by engelanna** [comment]
> Also, I had a bunch of refactor in the branch that you found that may be a good idea to pursue, but it's been neglected. What do you think of the API there? (I imagine it's very stale also)

I think the [cv_update_api](https://github.com/clintval/genome-browser/tree/cv_update_api) branch is better structured indeed :+1: Will base my PR off of that one, then - as soon as I've resolved the `xlim` max being clipped for some reason (mentioned in this PR's description).

This lib is exactly what I was looking for, being a relative plotting newbie who wanted something simple that just worked. Nice one :smile:  Would love to show off what I'm using it for, but there's still a couple of hurdles to jump across (1-2 weeks).

**Comment by engelanna** [comment]
Ah, to budge the [cv_update_api](https://github.com/clintval/genome-browser/tree/cv_update_api) branch, I'd need to know what the `targets` parameter to `GenomeDiagram.__init__` is:

```
class GenomeDiagram(object):
    def __init__(self, targets, name=None):
        self.targets = [targets] if not isinstance(targets, (list, tuple)) else targets
```

The example script is missing from the README there.

**Comment by clintval** [comment]
I think it was meant to be multiple genomic intervals so I could build in faceting. The first pass only looked at one interval at a time, so my dream was to easily support multiple intervals in one figure, perhaps by using faceting.

**Comment by engelanna** [comment]
@clintval Closing this in favour of https://github.com/clintval/genome-browser/pull/2.

Can't add you as a reviewer, no repo rights. Cheers! :clinking_glasses:

**Pull request closed without merging**